Logging guide (#13)
uasal/uasal_development_guide: source/python/python_logging_guide.rst

Python Logging Guide
====================

Written by Sanchit Sabhlok and Stephanie Rinaldi on 05/28/2025

Introduction
============

> This is an intermediate level guide, written for people who may not have experience in logging, but have some experience with python, object oriented programming and package development. This guide will discuss the `logger` module in python specifically, and the suggested practices for developing packages at UASAL. 

Author: done

Users are encouraged to look at the `Logging HOWTO <https://docs.python.org/3/howto/logging.html>`__ tutorial on the python website. If you have already been there and there is way too much information, and you would like a streamlined version of logging for your needs, this guide will reproduce some material from the tutorial, and add additional guidelines and suggestions specifically for UASAL package development.

Quickstart
==========

> To get started with logging, the user needs to add a logger class to their UASAL package.

Author: I can clarify the language here but I think this "Quickstart" is for people who are trying to add a standard UASAL logger to their packages and starting to use it, not necessarily as an "Intro to Logging" quickstart. I can clarify this.

> To get started with logging, the user needs to add a logger class to their UASAL package.

Reviewer: Yeah, it wasn't clear to me form the beginning that that was what the document was focusing on, so including / clarifying it in the statement of purpose at the beginning would help the reader. Otherwise it's good stuff!

> To get started with logging, the user needs to add a logger class to their UASAL package.

Author: Text and structure of the guide has been updated to reflect changes discussed here. The changes are not exactly on the spot changed lines or language, so I will resolve this thread for now.

> To get started with logging, the user needs to add a logger to their UASAL package/script. The basic logger usage is fairly straightforward and can be seen in `this gist <https://gist.github.com/sfrinaldi/df1f2711b70e209a161321231adf1d58>`__ ::

Reviewer 2: By the way / this is pointing to the test gist script that is using the logger class. Not sure if you meant to have that here still in this section or just later.

Author: I've left it in on purpose. There's a full exploration and a more toned down version in the codeblock

> 	import logging

Author: Added a clarification and link to the section.

> 	log = logging.getLogger(__name__)

Reviewer 2: I agree on that / makes sense. (Have a habit of shortening variables as much as possible without it being just a letter -> pretty sure Sanchit had copied that over from my examples / on me for that)

::

   ├── docs
   ├── notebooks
   │   ├── sample_notebook.ipynb
   ├── pyproject.toml
   ├── README.md 
   ├── src
   │   ├── package_name
   │   │   ├── source_code.py
   │   │   ├── __init__.py
   │   │   ├── packaged_directory
   │   │   │   └── data.csv
   │   │   └── _version.py
   └── tests
       └── test.py

Create a new file `logger.py` in the `src/package_name` directory. Inside this file, import the code from `the gist here <https://gist.github.com/sfrinaldi/ae2155aac8acbc6d6cc1ec750610022a>`__. Modify the default parameters to suit the needs of your package. Once this has been added, inside the `source_code.py` file, add ::

	from logger import logger
	
	class foo:
		def __init__(self):

> 			self.log = logger()

Reviewer 2: log = Logger(__name__)

Reviewer 2: This can just be called at the top of the script. If you use multiple loggers, you can set the name different for each. (logger-test.py gist for the example / test).

Author: updated

> Now this log can be used throughout the code to log messages, with formatting and handlers as prescribed in the class. Simply use the `self.log.debug()`, `self.log.info()`, `self.log.warn()`, `self.log.error()` and `self.log.critical()` methods to log the information accordingly.

Reviewer 2: Probably remove the `self` part in this now for the methods. It will be using just the base logger package methods / no need to add the self for this I would say now.

Author: I have removed this paragraph. Irina is right, after removing the logger class, the text is repetitive.

> Now this log can be used throughout the code to log messages, with formatting and handlers as prescribed in the class. Simply use the `self.log.debug()`, `self.log.info()`, `self.log.warn()`, `self.log.error()` and `self.log.critical()` methods to log the information accordingly.

Author: updated.

When importing the package, the log is automatically referenced as `foo.log` and all logging methods can be used with this object. 

Preamble
========

##############
Why not print?
##############

> A really simple question to ask is: Why do I want to use a logger and not simply print statements in my code? This is a reasonable question and the answer depends on your use case. If you are managing a relatively simple script, maybe a print statement is in fact the best way to go. But, if you are working with a complicated package, peppering the code with print statements to debug a problem has three main issues. First, you probably have to go back and comment out most statements as you will not want them once the code starts to work. Secondly, a print statement does not contain information about the line of code/date/time that triggered it, unless explicitly specified. Lastly, a large number of print statements is often redundant when debugging, since we are usually interested in the last few statements before something breaks, requiring the user to sift through a number of print statements to identify the important one.

Reviewer 3: 10000000% @stefi07

> A really simple question to ask is: Why do I want to use a logger and not simply print statements in my code? This is a reasonable question and the answer depends on your use case. If you are managing a relatively simple script, maybe a print statement is in fact the best way to go. But, if you are working with a complicated package, peppering the code with print statements to debug a problem has three main issues. First, you probably have to go back and comment out most statements as you will not want them once the code starts to work. Secondly, a print statement does not contain information about the line of code/date/time that triggered it, unless explicitly specified. Lastly, a large number of print statements is often redundant when debugging, since we are usually interested in the last few statements before something breaks, requiring the user to sift through a number of print statements to identify the important one.

Author: added

> A really simple question to ask is: Why do I want to use a logger and not simply print statements in my code? This is a reasonable question and the answer depends on your use case. If you are managing a relatively simple script, maybe a print statement is in fact the best way to go. But, if you are working with a complicated package, peppering the code with print statements to debug a problem has three main issues. There are logistical and scientific reasons for logging instead of printing. First, you probably have to go back and comment out most statements as you will not want them once the code starts to work. Secondly, a print statement does not contain information about the line of code/date/time that triggered it, unless explicitly specified. Lastly, a large number of print statements is often redundant when debugging, since we are usually interested in the last few statements before something breaks, requiring the user to sift through a number of print statements to identify the important one. Loggers are a strong way of circumventing these problems. Additionally, there is a strong reason from a scientific point of view to use logging as well for transparency, traceability and reproducibility of results. When logs are well-placed and informative they act as a lab-book for the data / results generated by the code.

Author: That reads better. Added.

5. Debug (Logger level numeric value - 10) - Detailed information, which the user almost never wants, and is typically useful for the code maintainers/designers. *Detailed information, typically of interest only when diagnosing problems.*

> A special note on the `log.warn()` keyword. Python has an additional package called `warnings` which also contains the `warnings.warn()` method, utilized frequently by packages. The distinction between the use of these packages is completely dependent on the package designer, BUT it is advisable to follow the following convention: *`warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning. A logger’s `log.warn()` method if there is nothing the client application can do about the situation, but the event should still be noted.*

Reviewer 2: `log.warn()` -> `log.warning()`

Now if we go back to our printing problem, we can instead replace them in the code with `logger.debug()` statements instead, which will only be printed when the logger itself is configured to print out debug messages. Once the problem is solved, we can leave the logging code in there, and simply switch the logger to an information only mode. No more commenting back and forth!

Intermediate/Advanced Usage
===========================

A standard logger workflow is illustrated in the following flowchart - 

|image1|

> While convenient to quickly use the logger defined above in a short script, when developing a package, it is useful to define a separate logger class to be used by the package with a number of pre-set default values that configure the logger.

Author: added

> A complete code snippet that can be copy-pasted into a package can be found `here <https://gist.github.com/sfrinaldi/ae2155aac8acbc6d6cc1ec750610022a>`__. The guide will now break this example down to discuss specific logger settings.

Author: Re reading the whole thing, I think the line is not required. Have removed this text.

	>>> log.debug("Testing log debugs")
	>>> log.warning("Testing log warnings")
	Testing log warnings
	>>> log.info("Testing log info")
	>>> log.error("Testing log errors")
	Testing log errors
	>>> log.critical("Testing log criticals")
	Testing log criticals

So the default behavior is to print everything above a warning to screen. If you haven't configured a FileHandler, none of this gets saved to a log file. You can change the output here by changing the level of your root logger. You can save all of this information to a file by configuring and adding a FileHandler.

Root Logger Level, Handler Level and throughput
===============================================

Loggers are effectively a chain of communication. The chain starts with the `logger` object, which then usually hands it off to one of the Handlers, or multiple handlers. The important thing to note is that *ALL OF THESE OBJECTS CAN SET THEIR OWN LEVEL*. In these circumstances, it is advisable to set the root logger's default level as Debug, so that all logged messages are passed on to the handlers. The handler levels can then be set to filter out a certain set of messages. For example, in this setup 
::

> 

Reviewer 2: I think we would actually want this to stay here for the most 'simple' version (or should I say a simplish version). (Whether we decide to keep the changes I have on the logger gists or not)

Python Logging Guide
====================

Introduction
============

> This is an intermediate level guide, written for people who may not have experience in logging, but have some experience with python, object oriented programming and package development. This guide will discuss the `logger` module in python specifically, and the suggested practices for implementing logging when developing packages at UASAL. 

Reviewer 4: python --> Python (numerous instances)

Author: fixed

	import logging
	logger = logging.getLogger(__name__)

	# Test log statements
	logger.debug("Testing log debugs")
	logger.warning("Testing log warnings")
	logger.info("Testing log info")
	logger.error("Testing log errors")
	logger.critical("Testing log criticals")

The workflow is to import a logger, and use the logger to log messages. The logger recognizes the pre defined set of levels, which are also function calls for specific messages. 

##############
Why not print?
##############

A really simple question to ask is: *Why do I want to use a logger and not simply print statements in my code?* 

This is a reasonable question and the answer depends on your use case. There are logistical and scientific reasons for logging instead of printing. If you are managing a relatively simple script, maybe a print statement is in fact the best way to go. But, if you are working with a complicated package, peppering the code with print statements to debug a problem has three main issues: 

1. Cleanup overhead - you probably have to go back and comment out most statements as you will not want them once the code starts to work. 

> 2. Lack of context - a print statement does not contain information about the line of code/date/time that triggered it, unless explicitly specified.

Reviewer 4: ```suggestion
2. Lack of context - a print statement does not contain information about the originating method, line of code, date, time that triggered it etc, unless explicitly specified.
```

Author: added

A special note on the `log.warning()` keyword. Python has an additional package called `warnings` which also contains the `warnings.warn()` method, utilized frequently by packages. The distinction between the use of these packages is completely dependent on the package designer, BUT it is advisable to follow the following convention: *`warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning. A logger’s `log.warning()` method if there is nothing the client application can do about the situation, but the event should still be noted.*

Now if we go back to our printing problem, we can instead replace them in the code with `logger.debug()` statements instead, which will only be printed when the logger itself is configured to print out debug messages. Once the problem is solved, we can leave the logging code in there, and simply switch the logger to an information only mode. No more commenting back and forth!

Intermediate/Advanced Usage
===========================

A standard logger workflow is illustrated in the following flowchart - 

|image1|

> While convenient to quickly use the logger defined above in a short script, when developing a package, it is useful to define a separate logger class to be used by the package with a number of pre-set default values that configure the logger. Here we outline the suggested practice for incorporating a custom logger into a UASAL package.

Reviewer 4: ```suggestion
While convenient to quickly use the logger defined above in a short script, when developing a package, one may find it useful to define a separate logger class to be used by the package with a number of pre-set default values that configure the logger. Here we outline the suggested practice if one decides to incorporate a custom logger into a UASAL package.
```

Author: added

A complete code snippet that can be copy-pasted into a package can be found `here <https://gist.github.com/sfrinaldi/ae2155aac8acbc6d6cc1ec750610022a>`__. 
::

   ├── docs
   ├── notebooks
   │   ├── sample_notebook.ipynb
   ├── pyproject.toml
   ├── README.md 
   ├── src
   │   ├── package_name
   │   │   ├── source_code.py
   │   │   ├── __init__.py
   │   │   ├── packaged_directory
   │   │   │   └── data.csv
   │   │   └── _version.py
   └── tests
       └── test.py

> Create a new file `logger.py` in the `src/package_name` directory. Inside this file, import the code from `the gist here <https://gist.github.com/sfrinaldi/ae2155aac8acbc6d6cc1ec750610022a>`__. Modify the default parameters to suit the needs of your package. Once this has been added, inside the `source_code.py` file, add ::

Reviewer 4: We really shouldn't be pointing people to anything on someone's private github repo. 
I think i'd rather see it in this repo than on a private one? Open to other ideas.

Author: @sfrinaldi can we move this gist to UASAL somehow?

Reviewer 2: ^@sanchitsabhlok Unfortunately (had this in the issue) there are no organization level gists / only can be created under a user profile. I could maybe in the future if its worth it have it so there is a simple workflow that just clones that gist into an appropriate repo every time its updated if there are other similar items for resources that would go into that repository as well. There isn't really a current repo that does that besides this one in a way but people wouldn't be normally looking around the repo itself I feel.

Author: Ok if we can't create a new gist, and we don't want to point to a private repo, the straightforward solution is create codeblocks at the bottom of the guide and change all links to point to those sections.

Reviewer 2: If there was a repo for where people could share this gists / have a workflow that pulls and updates from there, I would lean towards that over having a copy of this somewhere. Mainly like the gists feature (and wise it was available / implemented at the org level) since you can apply changes to each single file / easier for sharing code snippets without having people to worry or deal with updating their branches on a repo and merge conflicts if it was just pushing to a repo where everyone just uploaded their code snippets too.

Reviewer 2: Though I think that gist should be public (general note)

Author: Have applied the patch here for now. If we do add a repo in the future for gists, can always just come back and update this.

	from logger import logger
	
	class foo:
		def __init__(self):
			self.log = logger()

When importing a package that implements its own custom logger, the logger is automatically referenced as `foo.log` and all logging methods can be used with this object. 

> ######################

Reviewer 4: shouldn't this go above the custom logging class?

I think the custom class should go at the very bottom.

Author: moved and created a new section at the bottom.

	print("Running logger example with log_test setup w/mid line changes to logger values...")
	print("Console should only show message >='info' & record to log_test.log at 'crit' level")
	log_test.debug("Testing log debugs")
	log_test.warning("Testing log warnings")
	log_test.info("Testing log info")
	log_test.error("Testing log errors")
	log_test.critical("Testing log criticals")

	# Remove log testing statements / only need the import and initial log = Logger(__name__)
	# Other lines are for additional examples / test examples if running with logger.py locally

###################
Custom logger class
###################

Here we outline the suggested practice if one decides to incorporate a custom logger into a UASAL package. Given your package tree 
::

   ├── docs
   ├── notebooks
   │   ├── sample_notebook.ipynb
   ├── pyproject.toml
   ├── README.md 
   ├── src
   │   ├── package_name
   │   │   ├── source_code.py
   │   │   ├── __init__.py
   │   │   ├── packaged_directory
   │   │   │   └── data.csv
   │   │   └── _version.py
   └── tests
       └── test.py

Create a new file `logger.py` in the `src/package_name` directory. Inside this file, add the following code block. 

> ::

Reviewer 2: After that is clarified or adjusted think this is good to go / looks like the others are resolved.

> ::

Author: snippet above is just discussing handlers and is extracted from the class code. Minor duplication but to highlight the important points. The code below is the full version (meant for copy/pasting)

> ::

Author: Reduced line 146 to essentials.
